### PR TITLE
Added a method for returning the Decoder's current input encoding, and fix bugs

### PR DESCRIPTION
**Pull request by dralley**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -591,44 +600,18 @@ mod tests {
         s
     }
 
-    // In cases where all we have is a bom, we expect the bytes to be
-    // passed through unchanged.
-    #[test]
```
Comment (dralley) [inline src/lib.rs:596]: This requires some explanation.

The BOM is not random bytes, it is a unicode character, which is decodable just like any other unicode character. The bytes that comprise the BOM are merely the representation of that unicode character in a particular encoding.

You never, ever want to just "pass the bytes through unchanged".

File: src/util.rs [hunk 2]
```diff
@@ -209,9 +209,6 @@ impl PossibleBom {
     /// encoding. Otherwise, return `None`.
     pub fn encoding(&self) -> Option<&'static Encoding> {
         let bom = self.as_slice(true);
-        if bom.len() < 3 {
-            return None;
-        }
```
Comment (dralley) [inline src/util.rs:214]: The BOM is only 2 bytes long in UTF-16, so this always returned `None` in that case.

File: src/lib.rs (hunk 1 above)
Comment (BurntSushi) [reply src/lib.rs:596]: I understand everything you've written here, but don't understand why that has led you to remove the test. Could you say more please?

File: src/lib.rs [hunk 3]
```diff
@@ -410,6 +410,15 @@ impl<R: io::Read> DecodeReaderBytes<R, Vec<u8>> {
     pub fn new(rdr: R) -> DecodeReaderBytes<R, Vec<u8>> {
         DecodeReaderBytesBuilder::new().build(rdr)
     }
+
+    /// Return the encoding used by this decoder, which may have been
+    /// detected automatically or set manually.
+    ///
+    /// If automatic encoding detection is being used, the return value
+    /// of this method can change during the life of the decoder.
```
Comment (BurntSushi) [inline src/lib.rs:427]: This seems very non-ideal to me. How can one correctly use this method?

File: src/lib.rs [hunk 4]
```diff
@@ -132,7 +132,8 @@ impl DecodeReaderBytesBuilder {
         }
     }
 
-    /// Build a new decoder that wraps the given reader.
+    /// Build a new decoder that wraps the given reader. The default buffer
+    /// size is 8kb.
```
Comment (BurntSushi) [inline src/lib.rs:136]: The default buffer size is not an API guarantee and shouldn't be in the docs like this.

File: src/lib.rs [hunk 5]
```diff
@@ -555,10 +555,7 @@ impl<R: io::Read, B: AsMut<[u8]>> DecodeReaderBytes<R, B> {
                 self.buflen < self.buf.as_mut().len(),
                 "internal buffer should never be exhausted"
             );
-            let buf = self.buf.as_mut();
-            for (dst, src) in (self.pos..self.buflen).enumerate() {
-                buf[dst] = buf[src];
-            }
+            self.buf.as_mut().copy_within(self.pos..self.buflen, 0);
```
Comment (BurntSushi) [inline src/lib.rs:558]: Nice. I don't think this method existed when I wrote this code.

File: src/lib.rs (hunk 1 above)
Comment (dralley) [reply src/lib.rs:596]: The tests started failing as soon as I fixed the bug here https://github.com/BurntSushi/encoding_rs_io/pull/11#discussion_r962187963

It was returning the bytes unchanged was because it wasn't detecting an encoding. Once the encoding was detected, it starts sending bytes through the decoder, which strips the BOM because `encoding.new_decoder_with_bom_removal()` is used.  So the tests failed because the output was empty.

What probably ought to be done instead of just removing them entirely is - fix the decoding part as described, so that the BOM is not removed by the decoder, and then have these tests ensure that a BOM is returned in all cases where present in the input except when the user explicitly asked for its removal via `strip_bom`.

File: src/lib.rs (hunk 3 above)
Comment (dralley) [reply src/lib.rs:427]: The same way you would use it with `encoding_rs` https://docs.rs/encoding_rs/latest/encoding_rs/struct.Decoder.html#method.encoding

It's really only possible on the first read rather than at any random point, I can clarify that in the docs.  But this is partially why I prefer having a public `rdr.detect_encoding()` or  `rdr.detect_encoding_with(...)?;` method which would make the check explicit.

File: src/lib.rs (hunk 4 above)
Comment (dralley) [reply src/lib.rs:136]: I agree that it's not a guarantee but it's still worth mentioning even if we explicitly say it's not an API guarantee.  I can copy the language from `std::io::BufReader` if you like https://doc.rust-lang.org/std/io/struct.BufReader.html#method.new

File: src/lib.rs (hunk 3 above)
Comment (BurntSushi) [reply src/lib.rs:427]: I'm not convinced that's correct. Who is to say that the first read pulls in enough bytes for encoding detection to work? It is perfectly legal and correct for read calls to write at most one byte into the caller provided buffer.

File: src/lib.rs (hunk 1 above)
Comment (BurntSushi) [reply src/lib.rs:596]: I still don't get it. I'm going to have to go through and recontexualize the library impl in order to accept this. I don't know when I'll get the chance to do that.

File: src/lib.rs (hunk 3 above)
Comment (dralley) [reply src/lib.rs:427]: I believe that is the problem that [`utils::read_full`](https://github.com/BurntSushi/encoding_rs_io/blob/master/src/util.rs#L222-L243) exists to solve, do you think it would be sufficient to just use this function to fill the first buffer?

File: src/lib.rs (hunk 1 above)
Comment (dralley) [reply src/lib.rs:596]: Ok.  If it helps, I can walk through my own steps.

* Check out the first commit ([16b3203](https://github.com/BurntSushi/encoding_rs_io/pull/11/commits/16b320367f700061156361a56dc4ca6b94db3801)) and add these lines back https://github.com/BurntSushi/encoding_rs_io/pull/11#discussion_r962187963
* Observe that the new `PossibleBom::encoding()` tests fail by returning `None` when they ought to recognize a UTF-16 BOM and return `Some($encoding)`
* Review the tests for correctness.  Reset back to the original commit.  Observe that with the original commit, now the `trans_utf16_bom` tests are failing

```
---- tests::trans_utf16_bom stdout ----
thread 'tests::trans_utf16_bom' panicked at 'assertion failed: `(left == right)`
  left: `[255, 254]`,
 right: `[]`', src/lib.rs:602:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::trans_utf16_bom

test result: FAILED. 36 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

* After debugging it becomes clear that the reason was because the decoder was swallowing the BOM even though `DecodeReaderBytes` was not configured to strip the BOM.  And that is because we're creating the decoder with `encoding.new_decoder_with_bom_removal()`

File: src/lib.rs (hunk 4 above)
Comment (dralley) [reply src/lib.rs:136]: Obviously it's your call, but considering the constant in question is 2 lines away and the consistency with the standard library, I'm not sure I understand the big deal.

Comment (dralley) [reply src/lib.rs:136]: But if you're strongly against it, that's fine.

File: src/lib.rs (hunk 1 above)
Comment (dralley) [reply src/lib.rs:596]: The actual behavior it should probably expect, is not to get either an empty buffer or `[FF, FE]`, but to get `[0xEF, 0xBB, 0xBF]`, the BOM character decoded to UTF-8.

Comment (dralley) [reply src/lib.rs:596]: @BurntSushi As a practical matter, I need to know if this can possibly be merged soon or if I should just vendor the code.  Will you get an opportunity to review soon?

Comment (dralley) [reply src/lib.rs:596]: @BurntSushi ^

File: src/lib.rs (hunk 4 above)
Comment (BurntSushi) [reply src/lib.rs:136]: Please remove this change.

File: src/lib.rs [hunk 6]
```diff
@@ -352,6 +353,13 @@ impl DecodeReaderBytesBuilder {
 /// buffer used to store the results of transcoding. Callers may elect to reuse
 /// the internal buffer via the `DecodeReaderBytesBuilder::build_with_buffer`
 /// constructor.
+///
+/// Because there is an internal buffer, the inner type `R` does not need to be
+/// buffered separately via `BufReader` - that is, you can wrap types such as
```
Comment (BurntSushi) [inline src/lib.rs:358]: Is this supposed to be an em-dash? I think I would just use a period here and start a new sentence with `That is, ...`.

File: src/lib.rs (hunk 3 above)
Comment (BurntSushi) [reply src/lib.rs:427]: I would like to see a real example how of this method is being used, and ideally, it should be ported to a documentation test. I don't understand what this is supposed to be used for.

File: src/lib.rs (hunk 1 above)
Comment (BurntSushi) [reply src/lib.rs:596]: Is this change an extension of https://github.com/hsivonen/encoding_rs/issues/88? Because I generally agree with @hsivonen there. Otherwise, I'm still going to have to re-contextualize myself, and as I already said, **I don't know** when I'll be able to do this. Basically, what you're saying just doesn't make sense to me. That doesn't mean you're wrong. What I know is that this crate's implementation is rather subtle and I am just not okay with merging things I don't understand.

Basically, I wrote this test for _some_ reason, so if we're going to change the code in a way that results in the test failing, then I personally need to be exceptionally clear that it's indeed the right thing to do. And that it isn't going to break downstream folks. Including me. A lot of this crate was written for use in ripgrep for example.

**Review by BurntSushi — CHANGES_REQUESTED** [review]
